packages/api/src/middlewares/integration.ts

```diff
@@ -0,0 +1,92 @@
+import { TRPCError } from "@trpc/server";
+
+import { and, eq, inArray } from "@homarr/db";
+import { integrations } from "@homarr/db/schema/sqlite";
+import type { IntegrationKind } from "@homarr/definitions";
+import { z } from "@homarr/validation";
+
+import { decryptSecret } from "../router/integration";
+import { publicProcedure } from "../trpc";
+
+// TODO: add access control in some way, most likely reusable middleware or something
+
+export const createOneIntegrationMiddleware = <TKind extends IntegrationKind>(
+  ...kinds: TKind[]
+) => {
+  return publicProcedure
+    .input(z.object({ integrationId: z.string() }))
+    .use(async ({ input, ctx, next }) => {
+      const integration = await ctx.db.query.integrations.findFirst({
+        where: and(
+          eq(integrations.id, input.integrationId),
+          inArray(integrations.kind, kinds),
+        ),
+        with: {
+          secrets: true,
+        },
+      });
+
+      if (!integration) {
+        throw new TRPCError({
+          code: "NOT_FOUND",
+          message: "Integration not found",
+        });
+      }
+
+      const { secrets, kind, ...rest } = integration;
+
+      return next({
+        ctx: {
+          integration: {
+            ...rest,
+            kind: kind as TKind,
+            decryptedSecrets: secrets.map((secret) => ({
+              ...secret,
+              value: decryptSecret(secret.value),
+            })),
+          },
+        },
+      });
+    });
+};
+
+export const createManyIntegrationMiddleware = <TKind extends IntegrationKind>(
+  ...kinds: TKind[]
+) => {
+  return publicProcedure
+    .input(z.object({ integrationIds: z.array(z.string()) }))
```

Comment:
Can't you use zod to make it at least 1 ?

Reply:
yes with `.min(1)`